planning/autoware_path_generator/src/node.cpp

```diff
@@ -0,0 +1,418 @@
+    lanelet_sequence, *planner_data_.lanelet_map_ptr, params.waypoint_group_separation_threshold,
+    params.waypoint_group_interval_margin_ratio);
+
+  for (const auto & [waypoints, interval] : waypoint_groups) {
+    const auto overlap_start_it = std::find_if(
+      path_points.begin(), path_points.end(), [&](const lanelet::BasicPoint3d & point) {
+        return get_arc_length_along_centerline(lanelet_sequence, point) >= interval.first;
+      });
+    const auto overlap_end_it =
+      std::find_if(overlap_start_it, path_points.end(), [&](const lanelet::BasicPoint3d & point) {
+        return get_arc_length_along_centerline(lanelet_sequence, point) > interval.second;
+      });
+    if (overlap_start_it == overlap_end_it) {
+      continue;
+    }
+    const auto waypoint_start_it = path_points.erase(overlap_start_it, overlap_end_it);
+    path_points.insert(waypoint_start_it, waypoints.begin(), waypoints.end());
+  }
+
+  std::vector<PathPointWithLaneId> path_points_with_lane_id{};
+
+  const auto add_path_point = [&](const auto & path_point, const lanelet::ConstLanelet & lanelet) {
+    PathPointWithLaneId path_point_with_lane_id{};
+    path_point_with_lane_id.lane_ids.push_back(lanelet.id());
+    path_point_with_lane_id.point.pose.position =
+      lanelet::utils::conversion::toGeomMsgPt(path_point);
+    path_point_with_lane_id.point.longitudinal_velocity_mps =
+      planner_data_.traffic_rules_ptr->speedLimit(lanelet).speedLimit.value();
+    path_points_with_lane_id.push_back(std::move(path_point_with_lane_id));
+  };
+
+  for (auto it = path_points.begin(); it != path_points.end(); ++it) {
+    const auto & path_point = *it;
+
+    const auto lanelet_it = std::find_if(
+      lanelet_sequence.begin(), lanelet_sequence.end(), [&](const lanelet::ConstLanelet & lanelet) {
+        return get_arc_length_along_centerline(lanelet_sequence, path_point) >=
+               get_arc_length_along_centerline(lanelet_sequence, lanelet.centerline().front());
+      });
+    if (lanelet_it == lanelet_sequence.end()) {
+      continue;
+    }
+
+    const auto s = get_arc_length_along_centerline(lanelet_sequence, path_point);
+
+    if (s < s_start) {
+      if (it == std::prev(path_points.end())) {
+        break;
+      }
+      const auto & next_path_point = *std::next(it);
+      const auto s_next = get_arc_length_along_centerline(lanelet_sequence, next_path_point);
+      if (s_next > s_start) {
+        add_path_point(
+          get_interpolated_point(path_point, next_path_point, s_start - s), *lanelet_it);
+      }
+      continue;
+    }
+
+    if (s > s_end) {
+      if (it == path_points.begin()) {
+        break;
+      }
+      const auto & prev_path_point = *std::prev(it);
+      add_path_point(get_interpolated_point(path_point, prev_path_point, s - s_end), *lanelet_it);
+      break;
+    }
+
+    add_path_point(path_point, *lanelet_it);
+  }
+
+  utils::remove_overlapping_points(path_points_with_lane_id);
```

Comment:
will be unnecessary with Autoware Trajectory
https://autowarefoundation.github.io/autoware.universe/main/common/autoware_trajectory/